### PR TITLE
[CAD-779] Add a database backend.

### PR DESCRIPTION
**Pull request by ksaric**
https://jira.iohk.io/browse/CAD-779

### CODE REVIEW

File: src/DB.hs [hunk 1]
```diff
@@ -77,3 +101,39 @@ stubbedInitialDataMap = Map.fromList
...
+
+    , dlAddPoolMetadata     = \poolHash poolMetadata -> panic "To implement!"
+
+    , dlAddPoolMetadataSimple     = \poolHash poolMetadata -> do
+        let poolHashBytestring = (encodeUtf8 $ getPoolHash poolHash)
+        let poolEncodedMetadata = poolMetadata
+        let hashFromMetadata = B16.encode $ Crypto.digest (Proxy :: Proxy Crypto.Blake2b_256) $ (encodeUtf8 poolEncodedMetadata)
+
+        when (hashFromMetadata /= poolHashBytestring) $
+            panic "TxMetadataHashMismatch"
```
Comment (KtorZ) [inline src/DB.hs:132]: I get from this line that `poolHash` is actually not the pool id but the hash of the metadata? This feels rather confusing with the name "poolHash"?

File: src/Lib.hs [hunk 2]
```diff
@@ -140,10 +168,11 @@ postBlacklistPool user blacklistPool = convertIOToHandler $ do
     return examplePoolOfflineMetadata
 
 -- throwError err404
-getPoolOfflineMetadata :: PoolHash -> Handler PoolOfflineMetadata
-getPoolOfflineMetadata poolHash = convertIOToHandler $ do
+getPoolOfflineMetadata :: DataLayer -> PoolHash -> Handler PoolMetadataWrapped
+getPoolOfflineMetadata dataLayer poolHash = convertIOToHandler $ do
     putTextLn $ show poolHash
-    return examplePoolOfflineMetadata
+    fmap PoolMetadataWrapped $ either (\m -> panic $ renderLookupFail m) (\a -> a) <$> (dlGetPoolMetadataSimple dataLayer) poolHash
```
Comment (KtorZ) [inline src/Lib.hs:174]: As a result, this returns a _string_ which encodes a JSON object, and not a JSON object.

File: src/DB.hs (hunk 1 above)
Comment (ksaric) [reply src/DB.hs:132]: A lot of things in this PR are confusing, will address that.

File: src/Lib.hs (hunk 2 above)
Comment (ksaric) [reply src/Lib.hs:174]: Yes. Encoding it into a JSON object can change the JSON contents, and accordingly, the hash.

**Comment by ksaric** [comment]
Now returning the original JSON content on the correct hash (as agreed):
```
{"name": "test", "description": "This is a test pool", "ticker": "testy", "homepage": "https://github.com/input-output-hk/cardano-db-sync/tree/master/cardano-db/src/Cardano/Db"}
```

Returning the error structure (as agreed):
```
{"code":"DbLookupTxMetadataHash","description":"The hash cbdfc4f21feb0a414b2b9471fa56b0ebd312825e63db776d68cc3fa0ca1f5a2b is missing from the DB."}
```

File: src/Types.hs [hunk 3]
```diff
@@ -210,3 +220,37 @@ instance ToJSON PoolOfflineMetadata where
...
 
+
+newtype PoolMetadataWrapped = PoolMetadataWrapped Text
+    deriving (Eq, Ord, Show, Generic)
+
+-- Here we are usingg the unsafe encoding since we already have the JSON format
+-- from the database.
+instance ToJSON PoolMetadataWrapped where
+    toJSON (PoolMetadataWrapped metadata) = toJSON metadata
+    toEncoding (PoolMetadataWrapped metadata) = unsafeToEncoding $ encodeUtf8Builder metadata
```
Comment (KtorZ) [inline src/Types.hs:231]: This is still returning a plain string that contains encoded JSON. I could expect to be able to do:

```
curl -v http://localhost:3100/api/v1/metadata/{hash} > metadata.json
```

And end up with a valid JSON file. Now however, I need to decode it first! It is encoded to JSON twice!

```hs
λ> Right (String str) <- eitherDecodeFileStrict @Value "metadata.json" 
λ> Right metadata = eitherDecodeStrict @Value (T.encodeUtf8 str)
λ> metadata
Object (fromList [("homepage",String "https://github.com/input-output-hk/cardano-db-sync/tree/master/cardano-db/src/Cardano/Db"),("name",String "test"),("ticker",String "testy"),("description",String "This is a test pool")])
```

Comment (ksaric) [reply src/Types.hs:231]: Why can't you do `curl -v http://localhost:3100/api/v1/metadata/{hash} > metadata.json`?
When I do that, I end up with a valid JSON file.
```
$> curl -v http://localhost:3100/api/v1/metadata/cbdfc4f21feb0a414b2b9471fa56b0ebd312825e63db776d68cc3fa0ca1f5a2f > metadata.json

$> cat metadata.json

{"name": "test", "description": "This is a test pool", "ticker": "testy", "homepage": "https://github.com/input-output-hk/cardano-db-sync/tree/master/cardano-db/src/Cardano/Db"}
```

File: src/Cardano/Db/Error.hs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            ]
+
+
+renderLookupFail :: DBFail -> Text
+renderLookupFail lf =
+  case lf of
+    DbLookupTxMetadataHash hash -> "The hash " <> decodeUtf8 hash <> " is missing from the DB."
+    PoolMetadataHashMismatch -> "The pool metadata does not match!"
+    UnableToEncodePoolMetadataToJSON err -> "Unable to encode the content to JSON. " <> err
+    UnknownError text -> "Unknown error. Context: " <> text
```
[the remark this answers is not in the document]
Comment (ksaric) [reply src/Cardano/Db/Error.hs:68]: Good idea, will fix.

File: app/Main.hs [hunk 5]
```diff
@@ -4,5 +4,135 @@ import           Cardano.Prelude
...
+
+    -- Empty right now but we might add some params over time.
+    pInsertPool :: Parser Command
+    pInsertPool =
+      InsertPool <$> pFilePath <*> pPoolHash
+
+pFilePath :: Parser FilePath
+pFilePath =
+  Opt.strOption
+    (  Opt.long "filepath"
```
[the remark this answers is not in the document]
Comment (ksaric) [reply app/Main.hs:111]: That is fine, will fix.

File: src/Types.hs (hunk 3 above)
Comment (KtorZ) [reply src/Types.hs:231]: Hmm. Okay, stack trickery at play here, I was still running an old version. Re-compiling didn't help, I had to clear the `.stack-work` and re-compile from scratch. It indeed looks better! 

```console
$ curl -v http://localhost:3100/api/v1/metadata/cbdfc4f21feb0a414b2b9471fa56b0ebd312825e63db776d68cc3fa0ca1f5a2f  | jq
{
  "name": "test",
  "description": "This is a test pool",
  "ticker": "testy",
  "homepage": "https://github.com/input-output-hk/cardano-db-sync/tree/master/cardano-db/src/Cardano/Db"
}
```

**Comment by ksaric** [comment]
A note that might be of use - the argument name was changed from `filepath` to `metadata` when inserting the pool metadata:
```
PGPASSFILE=config/pgpass stack run smash-exe -- insert-pool --metadata test_pool.json --poolhash "cbdfc4f21feb0a414b2b9471fa56b0ebd312825e63db776d68cc3fa0ca1f5a2f"
```

File: README.md [hunk 6]
```diff
@@ -90,3 +90,69 @@ You need an HTTP server to serve it from and simply point to the application por
...
+```
+pg_dump -c -s --no-owner cexplorer > cexplorer.sql
+```
+
+## Inserting pool metadata
+
+
+This is an example (we got the hash from Blake2 256):
+```
+stack exec smash-exe -- insert-pool --filepath test_pool.json --poolhash "cbdfc4f21feb0a414b2b9471fa56b0ebd312825e63db776d68cc3fa0ca1f5a2f"
```
Comment (jbgi) [inline README.md:141]:
```suggestion
stack exec smash-exe -- insert-pool --metadata test_pool.json --poolhash "cbdfc4f21feb0a414b2b9471fa56b0ebd312825e63db776d68cc3fa0ca1f5a2f"
```

**Comment by ksaric** [comment]
@jbgi This was merged so you can base deployment commit from `master` if you want. New PR with some new changes should be going up today.

**Pull request merged**